src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/SampleRecordCollection.java

```diff
@@ -0,0 +1,173 @@
+package org.broadinstitute.hellbender.tools.copynumber.formats.collections;
+
+import com.google.common.collect.ImmutableList;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.copynumber.formats.metadata.SampleMetadata;
+import org.broadinstitute.hellbender.tools.copynumber.formats.metadata.SampleNameUtils;
+import org.broadinstitute.hellbender.tools.copynumber.formats.metadata.SimpleSampleMetadata;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.io.IOUtils;
+import org.broadinstitute.hellbender.utils.tsv.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Represents an immutable collection of records associated with a sample, a set of
+ * mandatory column headers given by a {@link TableColumnCollection}, and lambdas for
+ * reading and writing records.
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+public abstract class SampleRecordCollection<T> implements SampleMetadata {
+    private final SampleMetadata sampleMetadata;
+    private final ImmutableList<T> records;
+    private final TableColumnCollection mandatoryColumns;
+    private final Function<DataLine, T> dataLineToRecordFunction;
+    private final BiConsumer<T, DataLine> recordAndDataLineBiConsumer;
```

Comment:
How about `recordFromDataLineDecoder` and `recordToDataLineEncoder`?

As far as I can tell, there's no reason why the encoder for writing needs to be a BiConsumer, since you could start off each line with a fresh DataLine instance.  But for whatever reason, DataLine requires a line number, which is not really used anywhere.